Source/DafnyLanguageServer/Plugins/DafnyCodeActionHelpers.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System;
+using Microsoft.Boogie;
+using Microsoft.Dafny.LanguageServer.Language;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.Dafny.LanguageServer.Plugins;
+
+/// <summary>
+/// Helpers for plugins defining a DafnyCodeActionProvider
+/// </summary>
+public static class DafnyCodeActionHelpers {
+  /// <summary>
+  /// Given an LSP range and some text, extract the corresponding substring
+  /// </summary>
+  /// <param name="range">The range to extract</param>
+  /// <param name="text">The document</param>
+  /// <returns>The substring of the document in the range</returns>
+  public static string Extract(Range range, string text) {
+    var token = range.ToToken(text);
+    var startTokenPos = token.StartToken.pos;
+    var endTokenPos = token.EndToken.pos + token.EndToken.val.Length;
+    var length = endTokenPos - startTokenPos;
+    if (startTokenPos < 0 || endTokenPos < startTokenPos || endTokenPos >= text.Length) {
+      return ""; // Safeguard
+    }
+
+    return text.Substring(startTokenPos, length);
+  }
+
+  /// <summary>
+  /// Given the position of a closing brace '}' and the position of the opening brace '{',
+  /// returns spacing that can be used to insert a statement before the closing brace,
+  /// as well as spacing for after the statement
+  ///
+  /// For example, given:
+  /// ```
+  ///     {
+  ///       Some code here:
+  ///     }
+  /// ```
+  /// the method will return ("  ", "    "), so that it someone inserted the first value,
+  /// "assert X;\n" and the second value, the resulting code would be
+  /// ```
+  ///     {
+  ///       Some code here:
+  ///       assert X;
+  ///     }
+  /// ```
+  /// 
+  /// </summary>
+  /// <param name="endToken">The position of the closing brace</param>
+  /// <param name="text">The document text</param>
+  /// <returns>(extra indentation for a statement, current indentation)</returns>
+  public static (string, string) GetIndentationBefore(IToken endToken, int startLine, int startCol, string text) {
```

Comment:
In the long run, I imagine we might want to provide a different API for creating refactorings in Dafny, one that doesn't require thinking about formatting.

Here's a bad example:

```c#
Edit AddStatementToEnd(method: DafnyMethod) {
  var original = method.Snapshot();
  method.Body = method.Body.Concat(new SomeStatement(..));
  Formatter.Format(method);
  return DiffEngine.ComputeEdit(original, method);
}
```

I'm sure Roslyn has some interesting APIs for this very same purpose. I'd be great if we could get familiar with them, although I imagine it's quite a learning curve.

Here's an example: https://github.com/dotnet/roslyn/blob/main/src/Analyzers/CSharp/CodeFixes/MisplacedUsingDirectives/MisplacedUsingDirectivesCodeFixProvider.cs

With test: https://github.com/dotnet/roslyn/blob/main/src/Analyzers/CSharp/Tests/MisplacedUsingDirectives/MisplacedUsingDirectivesCodeFixProviderTests.cs

Reply:
Yes yes ! That's also part of what I envision we will be able to do with [the Dafny formater](https://github.com/dafny-lang/dafny/pull/2399).
In any case, in the short-term after this PR, we will need to handle expressions, substitutions, etc. So formatting them might be a good idea.

Reply:
Ok,

Just want us and our users to be aware that the API might still change (a lot)

Reply:
Yeah, I hope we can move toward AST-based rewrites relatively soon. All this string processing makes me nervous.